File: zirudemo/lib/clientengine/ClientEngine.dart
import 'dart:convert';
import 'dart:io';

import 'package:device_info/device_info.dart';
import 'package:flutter/material.dart';
import 'package:fluttertoast/fluttertoast.dart';
import 'package:path_provider/path_provider.dart';
import 'package:zirudemo/ZRConstants.dart';
import 'package:zirudemo/clientengine/form/ZiRuWebForm.dart';
import 'package:zirudemo/manager/FormListManager.dart';
import 'package:zirudemo/manager/FormViewManager.dart';
import 'package:zirudemo/manager/NetworkManager.dart';
import 'package:zirudemo/manager/ShareDataManager.dart';
import 'package:zirudemo/manager/SpManager.dart';
import 'package:package_info/package_info.dart';
import 'package:zr_flutter_native_bridge/zr_flutter_native_bridge.dart';

class ClientEngine {
  ///单例
  factory ClientEngine() => _getInstance();
  static ClientEngine get instance => _getInstance();
  static ClientEngine _clientEngine;
  ClientEngine._internal() {
    print("clientEngine 初始化 _internal");
  }
  static ClientEngine _getInstance() {
    if (_clientEngine == null) {
      _clientEngine = new ClientEngine._internal();
    }
    return _clientEngine;
  }

  SpManager _spManager; //数据持久化存储
  ShareDataManager _shareDataManager; // 数据内存保存
  ///包名 版本号 APP名称
  PackageInfo _packageInfo;

  IosDeviceInfo _iosDeviceInfo;
  AndroidDeviceInfo _androidDeviceInfo;
  FormViewManager formViewManager = FormViewManager();

  Future<bool> init() async {
    //初始化SP
    _spManager = new SpManager();
    if (!await _spManager.init()) {
      return false;
    }
    //初始化 ShareData
    _shareDataManager = new ShareDataManager();

    // 初始化文件路劲
    if (!await _iniFilePath()) {
      return false;
    }

    _packageInfo = await PackageInfo.fromPlatform();

    if (Platform.isIOS) {
      _iosDeviceInfo = await DeviceInfoPlugin().iosInfo;
    } else if (Platform.isAndroid) {
      _androidDeviceInfo = await DeviceInfoPlugin().androidInfo;
    }

    ///初始化flutter与原生插件之间的桥梁
    ZrFlutterNativeBridge();

    print("clientengine初始化完成");
    return true;
  }

  Future<bool> _iniFilePath() async {
    if (Platform.isIOS) {
      Directory exD = await getLibraryDirectory();
      ZRConstants.BasePath = exD.path;
      ZRConstants.zipPath = ZRConstants.BasePath + "/ziru/zip/";
    } else {
      if (ZRConstants.isDebug) {
        Directory exD = await getExternalStorageDirectory();
        ZRConstants.BasePath = exD.path;
      } else {
        Directory appDocDir = await getApplicationDocumentsDirectory();
        ZRConstants.BasePath = appDocDir.path;
      }
      ZRConstants.zipPath = ZRConstants.BasePath + "/ziru/zip/";
    }

    return true;
  }
//////////////////////////////////////////////////////////////////

  setSharedData(String strKey, String strValue) {
    _shareDataManager.setSharedData(strKey, strValue);
  }

  getSharedData(String strKey) {
    return _shareDataManager.getSharedData(strKey);
  }

  deleteSharedData(String strKey) {
    _shareDataManager.deleteSharedData(strKey);
  }

  saveData(String strKey, String strData) {
    _spManager.saveString(strKey, strData);
  }

  loadData(String strKey) {
    return _spManager.getString(strKey);
  }

  deleteData(String strKey) {
    _spManager.deleteString(strKey);
  }

  info(String message) {
    Fluttertoast.showToast(
        msg: message,
        toastLength: Toast.LENGTH_LONG,
        gravity: ToastGravity.CENTER,
        timeInSecForIos: 1,
        backgroundColor: Colors.red,
        textColor: Colors.white,
        fontSize: 16.0);
  }

  openform(BuildContext context, String url, String strTitle, String strData,
      int nOpenMode, int nAnimation) {
    String strOrginUrl = url;
    formViewManager.openform(
        context, strOrginUrl, strTitle, strData, nOpenMode, nAnimation);
  }

  back(BuildContext context, int nAnimation) {
    formViewManager.back(context, nAnimation);
  }

  ///异步startsdk
  startSDK(String callBackFuncName, String classPath, String methodName,
      String params) {
    print("startSDK : $callBackFuncName, $classPath, $methodName, $params");

    String urlMapping = this.loadData("urlmapping");

    // assert(urlMapping != null);

    String sdkName = classPath.replaceAll("class://", "");

    ZrFlutterNativeBridge.startSDK(sdkName, methodName, params).then((result) {
      print("异步获得的数据 = $result");
    });
  }

  ///同步startsynsdk
  Future<String> startSynSDK(String callBackFuncName, String classPath,
      String methodName, String params) async {
    print("startSynSDK : $callBackFuncName, $classPath, $methodName, $params");

    String urlMapping = this.loadData("urlmapping");

    // assert(urlMapping != null);

    String sdkName = classPath.replaceAll("class://", "");

    return "";
  }

  ///getSystemData 获取平台和APP信息
  getSystemData() {
    if (Platform.isIOS) {
      var array = [
        "2",
        _packageInfo.version.replaceAll(".", ""),
        _packageInfo.version,
        _iosDeviceInfo.systemVersion
      ];
      return array.join(',');
    } else if (Platform.isAndroid) {
      return [
        "1",
        _packageInfo.version.replaceAll(".", ""),
        _packageInfo.version,
        _androidDeviceInfo.version
      ];
    }
  }







  ///网络相关
  
  /*
  @param strCallbackFunctionName  js回调函数
  @param strRequest   cdo请求
  @param strAttachData    参数
  @param strUnCancel  可选参数   默认是0 传递1为不可取消
  */
  raiseTrans(String strCallbackFunctionName, 
             String strRequest, 
             String strAttachData, 
             String strUnCancel){
    
    final intUnCancel = strUnCancel != null ? int.parse(strUnCancel): 0;

  }

  Future<String> raiseHttpGet(String strURL) async{
    
    return NetworkManager().requestData(strURL, "1", null, null);
    
  }

  Future<String> raiseHttpPost(String strURL,
                               String strContentType,
                               String strData) async{
    var dataObj = jsonDecode(strData);
    return NetworkManager().requestData(strURL, "2", dataObj,strContentType);
  }


}

File: zirudemo/lib/html/ClientEngine.js
function ClientEngine() {


    /*
	 * Web窗体加载完毕，触发该事件
	 */
    // 客户端调用此方法
    this.onClientFormLoaded = function () {

        //_thisForm_.onClientFormLoaded();
        this.log("==onClientFormLoaded==");
    };

	/*
	 * 窗体得到输入焦点到前台时触发该事件
	 */
    this.onClientFormFocused = function () {
        //_thisForm_.onClientFormFocused();
        this.log("==onClientFormFocused==");
    };
	/*
	 * 窗体失去输入焦点到前台时触发该事件
	 */
    this.onClientFormUnfocused = function () {
        this.log("==onClientFormUnfocused==");
        //_thisForm_.onClientFormUnfocused();
    };

	/*
	 * Web窗口销毁
	 */
    this.onClientFormUnloaded = function () {
        this.log("==onClientFormUnloaded==");

        //_thisForm_.onClientFormUnloaded();
    };
	/*
	按下物理返回键事件
	*/
    this.onClientBackPressed = function () {
        //_thisForm_.onClientBackPressed();
        this.back(2);
    };
    this.openForm = function (strURI, strTitle, nOpenMode, strData, nAnimation) {
        var data = {};
        data["method"] = "openform";
        var params = {};
        params["url"] = strURI;
        params["nAnimation"] = nAnimation;
        params["strTitle"] = strTitle;
        params["nOpenMode"] = nOpenMode;
        params["strData"] = strData;
        data["params"] = params;
        clientJsEngine.postMessage(JSON.stringify(data))
    };
    this.back = function (nAnimation) {
        var data = {};
        data["method"] = "back";
        var params = {};
        params["nAnimation"] = nAnimation;
        data["params"] = params;
        clientJsEngine.postMessage(JSON.stringify(data))
    }
    /*
         退出应用
        */
    this.exitApp = function () {

    };

    this.setShareData = function (strKey, strValue) {
        var params = {};
        params['strKey'] = strKey;
        params['strValue'] = strValue;
        var resultjson = prompt(JSON.stringify(['zrJsEngine', 'setShareData', JSON.stringify(params)]));
        return resultjson;
    }

    this.getShareData = function (strKey) {
        var params = {};
        params['strKey'] = strKey;
        var resultjson = prompt(JSON.stringify(['zrJsEngine', 'getShareData', JSON.stringify(params)]));
        return resultjson;
    }
    this.deleteSharedData = function (strKey) {
        var params = {};
        params['strKey'] = strKey;
        var resultjson = prompt(JSON.stringify(['zrJsEngine', 'deleteSharedData', JSON.stringify(params)]));
        return resultjson;
    }

    /*
	拨打电话
	  @strPhoneNumber 电话号码
	*/
    this.dial = function (strPhoneNumber) {

    }
    this.getLocation = function () {

    }
    this.openBrowser = function (strUrl) {

    }
    /*
	保存数据项到存储（持久化）
	@strKey 数据的key
	@strData 数据的内容
	*/
    this.saveData = function (strKey, strValue) {

        var params = {};
        params['strKey'] = strKey;
        params['strValue'] = strValue;
        var resultjson = prompt(JSON.stringify(['zrJsEngine', 'saveData', JSON.stringify(params)]));
        return resultjson;
    }
    /*
      从存储中加载指定数据项的内容，如果不存在则返回null
   @strKey 数据项的key
*/
    this.loadData = function (strKey) {
        var params = {};
        params['strKey'] = strKey;
        var resultjson = prompt(JSON.stringify(['zrJsEngine', 'loadData', JSON.stringify(params)]));
        return resultjson;
    }
    /*
	      从存储中删除指定的数据项
	   @strKey 数据项的key，key不存在不影响
	*/
    this.deleteData = function (strKey) {
        var params = {};
        params['strKey'] = strKey;
        var resultjson = prompt(JSON.stringify(['zrJsEngine', 'deleteData', JSON.stringify(params)]));
        return resultjson;
    }
    /*
	/*
	     返回系统数据，整数数组字符串格式，比如[0,2000035]，其中
	    下标0为操作系统类型:1-安卓，2-IOS
	    下标1为框架版本号
	*/
    this.getSystemData = function () {

    }

    /*
     打开AppStore对应的App界面
*/
    this.openAppStore = function () {

    }
    /*
     返回设备码
*/
    this.getDeviceCode = function () {
        var params = {};
        var resultjson = prompt(JSON.stringify(['zrJsEngine', 'getDeviceCode', JSON.stringify(params)]));
        return resultjson;
    }
    /**
	 * 显示一个文本消息和确定按钮的模态对话框,显示信息3秒后消失
	 * @param  strText     弹出的消息内容
	 * @return
	 * @remark
	 */
    this.info = function (strText) {
        var data = {};
        data["method"] = "info";
        var params = {};
        params["message"] = strText;
        data["params"] = params;
        clientJsEngine.postMessage(JSON.stringify(data))
    }
    /*
	向服务器发起一个HttpGet请求调用，调用结果通过onHttpResult函数接收;同一个请求，在收到回调之前，再次调用忽略
	 @cdoRequest 请求对象
	 @strCallbackFunctionName 处理回调的函数名称，函数定义为XXX(nResponseCode,strResponseText)
	 @strAttachData 回调时携带上此数据
	*/
    this.raiseHttpGet = function (strCallbackFunctionName, strURL, strAttachData) {

    }
    /*
	   向服务器发起一个HttpPost请求调用，调用结果通过onHttpResult函数接收;同一个请求，在收到回调之前，再次调用忽略
	 @strCallbackFunctionName 处理回调的函数名称，函数定义为XXX(nResponseCode,strResponseText,strAttachData)
	 @strURL 请求URL
	 @strData 通过Post提交的文本数据
	 strContentType  json格式  
	 @strAttachData 回调时携带上此数据
	*/
    this.raiseHttpPost = function (strCallbackFunctionName, strURL, strContentType, strData, strAttachData) {

    }


    /**
    * @param strCallBackFunctionName 回调方法名称
    * @param strClassPath  类名              
    * @param strMethodName  方法名
    * @param params   参数 json字符串
    */
    this.startSDK = function (strCallBackFunctionName, strClassPath, strMethodName, strParams) {

    };
    /**
    * @param strClassPath  类名              
    * @param strMethodName  方法名
    * @param params   参数 json字符串
    */
    this.startSynSDK = function (strClassPath, strMethodName, strParams) {
        return clientEngine.startSynSDK(strClassPath, strMethodName, strParams);
    };
    this.log = function (value) {
        var data = {};
        data["method"] = "log";
        var params = {};
        params["value"] = value;
        data["params"] = params;
        clientJsEngine.postMessage(JSON.stringify(data))
    }


}
var clientEngine = new ClientEngine();